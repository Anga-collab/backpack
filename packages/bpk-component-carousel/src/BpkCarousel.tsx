import { useRef, useState } from 'react';


import { useScrollToInitialImage } from './utils';
import BpkCarouselContainer from './BpkCarouselContainer';
import BpkSlidesIndicator from './BpkCarouselSlidesIndicator';
import type { Props } from './types';

const BpkCarousel = ({
  images,
  initialImageIndex = 0,
  onImageChanged = null
}: Props) => {
  const [shownImageIndex, updateShownImageIndex] = useState(initialImageIndex);
  const imagesRef = useRef<Array<HTMLElement | null>>([]);

  useScrollToInitialImage(initialImageIndex, imagesRef);

  return (
    <>
        <BpkCarouselContainer
          images={images}
          onVisible={updateShownImageIndex}
          imagesRef={imagesRef}
          onImageChanged={onImageChanged}
        />
      <BpkSlidesIndicator
        length={images.length}
        activeIndex={shownImageIndex}
      />
      </>
  );
};

export default BpkCarousel;
