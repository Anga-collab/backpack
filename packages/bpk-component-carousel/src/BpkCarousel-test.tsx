import { render, screen, getAllByRole } from '@testing-library/react';
import { axe } from 'jest-axe';
import userEvent from '@testing-library/user-event';
import type { ReactNode } from 'react';

import BpkCarousel from './BpkCarousel';


describe('BpkCarousel', () => {
  let images: ReactNode[]

  const DemoImages = () => (
    <img src="http://content.skyscnr.com/available/949043373/949043373_343x132.jpg" alt='hotel bedroom' />
  )

  function generateDemoImages(count: number) {
    return Array.from({ length: count }, (_, i) => (<DemoImages />));
  }

  type TestCase = [
    expectedCount: 7 | 5 | 4 | 12,
    actualCount: number,
    props: React.ComponentProps<typeof BpkCarousel>,
  ];

  beforeAll(() => {

    window.HTMLElement.prototype.scrollIntoView = jest.fn();

    (window as any).IntersectionObserver = class IntersectionObserver {
      observe = jest.fn();

      disconnect = jest.fn();

      unobserve = jest.fn();
    };
    images = [<DemoImages />, <DemoImages />, <DemoImages />, <DemoImages />, <DemoImages />];
  });

  it('should render correctly', () => {
    const { asFragment } = render(<BpkCarousel images={images} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it.each<TestCase>([
    [12, 10, { images: generateDemoImages(10) }],
    [7, 5, { images: generateDemoImages(5) }],
    [5, 3, { images: generateDemoImages(3) }],
    [4, 2, { images: generateDemoImages(2) }],
  ])(
    'renders %i image(s) when there are %i image(s) available',
    (expectedCount, actualCount, props) => {
      render(<BpkCarousel {...props} />);

      expect(screen.getByTestId('image-gallery-scroll-container').childElementCount).toBe(expectedCount);
    },
  );

  it('renders only one image when only one available (no fake images for the infinite scroll)', async () => {
    render(<BpkCarousel images={[images[0]]} />);

    expect(screen.getByTestId('image-gallery-scroll-container').childElementCount).toBe(1);
  });
});
