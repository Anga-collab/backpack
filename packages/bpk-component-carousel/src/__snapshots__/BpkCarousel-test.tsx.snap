// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BpkCarousel should render correctly 1`] = `
<DocumentFragment>
  <div
    class="bpk-carousel-container"
    data-testid="image-gallery-scroll-container"
    role="list"
  >
    <div
      class="bpk-carousel-image"
      data-index="4"
      role="listitem"
    >
      <img
        alt=""
        src="https://d2xf5gjipzd8cd.cloudfront.net/available/949043373/949043373_343x132.jpg"
      />
    </div>
    <div
      class="bpk-carousel-image"
      data-index="0"
      role="listitem"
    >
      <img
        alt=""
        src="https://d2xf5gjipzd8cd.cloudfront.net/available/949043373/949043373_343x132.jpg"
      />
    </div>
    <div
      class="bpk-carousel-image"
      data-index="1"
      role="listitem"
    >
      <img
        alt=""
        src="https://d2xf5gjipzd8cd.cloudfront.net/available/949043373/949043373_343x132.jpg"
      />
    </div>
    <div
      class="bpk-carousel-image"
      data-index="2"
      role="listitem"
    >
      <img
        alt=""
        src="https://d2xf5gjipzd8cd.cloudfront.net/available/949043373/949043373_343x132.jpg"
      />
    </div>
    <div
      class="bpk-carousel-image"
      data-index="3"
      role="listitem"
    >
      <img
        alt=""
        src="https://d2xf5gjipzd8cd.cloudfront.net/available/949043373/949043373_343x132.jpg"
      />
    </div>
    <div
      class="bpk-carousel-image"
      data-index="4"
      role="listitem"
    >
      <img
        alt=""
        src="https://d2xf5gjipzd8cd.cloudfront.net/available/949043373/949043373_343x132.jpg"
      />
    </div>
    <div
      class="bpk-carousel-image"
      data-index="0"
      role="listitem"
    >
      <img
        alt=""
        src="https://d2xf5gjipzd8cd.cloudfront.net/available/949043373/949043373_343x132.jpg"
      />
    </div>
  </div>
  <div
    class="bpk-carousel-slides-indicator"
  >
    <div
      class="bpk-carousel-slides-indicator__row"
      style="--scroll-index: 0;"
    >
      <div
        class="bpk-carousel-slides-indicator__dot  bpk-carousel-slides-indicator__active"
      />
      <div
        class="bpk-carousel-slides-indicator__dot bpk-carousel-slides-indicator__siblingToActive "
      />
      <div
        class="bpk-carousel-slides-indicator__dot  "
      />
      <div
        class="bpk-carousel-slides-indicator__dot  "
      />
      <div
        class="bpk-carousel-slides-indicator__dot  "
      />
    </div>
  </div>
</DocumentFragment>
`;
